### PR TITLE
Refactor and extract inbox panel components

### PR DESCRIPTION
**Pull request by louwie17**
Needed for https://github.com/Automattic/woocommerce-payments/issues/1662

The big changes in this PR:
- Refactored the inbox note and action button component to be more modular and less opinionated (I pushed several things up stream by the use of callbacks).
- Refactored the components to functional components and into TypeScript
- I moved most of the inbox panel components to the experimental package so it could be used for the issue linked above in WC Pay

### Detailed test instructions:

- Load this branch and go to **WooCommerce > Home**
- Finish the onboarding flow if on a fresh install
- Smoke test the inbox notes
     - Dismiss a single note
     - Dismiss all the notes
     - Check if action buttons work correctly
     - Check if the `wcadmin_inbox_note_view`, `wcadmin_inbox_action_dismiss` tracks are triggered.

### CODE REVIEW

File: client/inbox-panel/index.js [hunk 1]
```diff
@@ -85,6 +170,103 @@ const InboxPanel = ( props ) => {
...
+
+		recordEvent( 'inbox_action_dismiss', {
+			note_name: dismiss.note.name,
+			note_title: dismiss.note.title,
+			note_name_dismiss_all: noteNameDismissAll,
+			note_name_dismiss_confirmation: confirmed,
+			screen,
+		} );
+
+		if ( confirmed ) {
```
Comment (louwie17) [inline client/inbox-panel/index.js:189]: This used to be in the note action component

File: packages/experimental/src/inbox-note/action.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
+/**
```
Comment (louwie17) [inline packages/experimental/src/inbox-note/action.tsx:1]: This component used to contain a lot of extra logic for removing notes, I removed all of that, and made it very dumb... 😆

File: packages/experimental/src/inbox-note/style.scss [hunk 3]
```diff
@@ -174,22 +174,6 @@
 	}
 }
 
-.woocommerce-homepage-notes-wrapper {
```
Comment (louwie17) [inline packages/experimental/src/inbox-note/style.scss:177]: Removed css unrelated to the inbox note components and moved back to the inbox-panel/index.scss

File: client/inbox-panel/index.js [hunk 4]
```diff
@@ -85,6 +170,103 @@ const InboxPanel = ( props ) => {
...
+			const noteId = dismiss.note.id;
+			if ( noteId && ! noteNameDismissAll ) {
+				removeNote( noteId )
```
Comment (samueljseay) [inline client/inbox-panel/index.js:192]: Yeesh this indentation is... blech..

Could we use async/await here?

File: client/inbox-panel/index.js [hunk 5]
```diff
@@ -85,6 +170,103 @@ const InboxPanel = ( props ) => {
...
+					.then( ( notesRemoved ) => {
+						createNotice(
+							'success',
+							__( 'All messages dismissed', 'woocommerce-admin' ),
+							{
+								actions: [
+									{
+										label: __(
+											'Undo',
+											'woocommerce-admin'
```
Comment (samueljseay) [inline client/inbox-panel/index.js:234]: Same here, I'm wondering if async/await would be a lot nicer.

File: packages/experimental/src/inbox-note/inbox-note.tsx [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	onNoteActionClick,
+	onBodyLinkClick,
+	onNoteVisible,
+} ) => {
+	const [ clickedActionText, setClickedActionText ] = useState( false );
+	const hasBeenSeen = useRef( false );
+	const toggleButtonRef = useRef< HTMLButtonElement >( null );
+
+	const handleBodyClick = ( event: {
+		target: EventTarget | HTMLBaseElement;
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/inbox-note.tsx:76]: I don't think this type is correct? I think you can just type the whole arg as: `React.MouseEvent< HTMLElement >`. and I *think* the target is just typed as `EventTarget`?

File: packages/experimental/src/inbox-note/inbox-note.tsx [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+	} ) => {
+		if ( 'href' in event.target ) {
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/inbox-note.tsx:78]: Is this ever relevant? it seems this handler is just used in the span `onClick`?

Comment (samueljseay) [reply packages/experimental/src/inbox-note/inbox-note.tsx:78]: It seems most of this code is just extracted, so I understand if you don't want to address non TS suggestions here.

File: packages/experimental/src/inbox-note/inbox-note.tsx [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+		if ( ! noteActions ) {
+			return;
+		}
+
+		return (
+			<>
+				{ noteActions.map( ( action, index ) => (
+					<InboxNoteActionButton
+						key={ index }
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/inbox-note.tsx:217]: We shouldn't use the `index` as a key, it leads to strange bugs, and as a result is just a general anti-pattern. Again I know this is probably just extracted code so up to you if you fix.

Probably a stable ID could be generated from some combo of label and url or something.

File: packages/experimental/src/inbox-note/test/inbox-dismiss-confirmation-modal.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import { render } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+
+/**
+ * Internal dependencies
+ */
+import { InboxDismissConfirmationModal } from '../inbox-dismiss-confirmation-modal';
+
+describe( 'InboxDismissConfirmationModal', () => {
+	it( 'should render with default button label - Yes, Iam sure', () => {
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/test/inbox-dismiss-confirmation-modal.tsx:13]: Nit pick but *Yes, I'm sure

**Review by samueljseay — COMMENTED** [review]
Looks good, just some minor suggestions from me.

File: packages/experimental/src/inbox-note/inbox-note.tsx (hunk 7 above)
Comment (louwie17) [reply packages/experimental/src/inbox-note/inbox-note.tsx:78]: It is suppose to still log a link click if a user clicks a link within the paragraph.
Unfortunately TypeScript does not like any of this, EventTarget has like no props so `href` always seems to be undefined, which is why I did the above.

Comment (louwie17) [reply packages/experimental/src/inbox-note/inbox-note.tsx]: I took this logic out and put it in a custom hook to make it more explicit: https://github.com/woocommerce/woocommerce-admin/pull/7006/commits/c0e877dbd77cea15729392028aefb2193c5db7f1

File: client/inbox-panel/index.js (hunk 4 above)
Comment (louwie17) [reply client/inbox-panel/index.js]: I refactored it a bit here: https://github.com/woocommerce/woocommerce-admin/pull/7006/commits/49820bb5a009bd74ccbd8a26e2d9f18c56d8d47b I did use async/away, which only really took away one indentation, I streamlined it a bit more, so it is less lengthy.

**Comment by louwie17** [comment]
@samueljseay Would you be able to have another look at this? I addressed most of your feedback, some directly, and some indirectly haha

File: packages/experimental/src/inbox-note/action.tsx [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+ * automatically open it in a new tab/window.
+ */
+const InboxNoteAction: React.FC< InboxNoteActionProps > = ( {
+	label,
+	onClick,
+	href,
+} ) => {
+	const [ inAction, setInAction ] = useState( false );
+
+	const handleActionClick: React.MouseEventHandler = ( event ) => {
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/action.tsx:26]: If you use the generic arg here you can combine it with `currentTarget` and remove the type narrowing. 

```suggestion
	const handleActionClick: React.MouseEventHandler<HTMLAnchorElement> = ( event ) => {
```

File: packages/experimental/src/inbox-note/action.tsx [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+		const target: EventTarget = event.target;
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/action.tsx:27]: Also note you don't need to specify types when they're inferred from by TS, here I change this to `currentTarget` which you'll find has the `href` available to it.

```suggestion
		const target = event.currentTarget;
```

File: packages/experimental/src/inbox-note/action.tsx [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+			disabled={ inAction }
+			href={ href }
+			onClick={ handleActionClick }
+		>
+			{ label }
+		</Button>
+	);
+};
+
+export default InboxNoteAction;
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/action.tsx:54]: Can we please avoid exporting defaults, it makes mass refactor's much harder due if anyone mis-names an import

File: packages/experimental/src/inbox-note/inbox-dismiss-confirmation-modal.tsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+					>
+						{ buttonLabel }
+					</Button>
+				</div>
+			</div>
+		</Modal>
+	);
+};
+
+export { InboxDismissConfirmationModal };
```
Comment (samueljseay) [inline packages/experimental/src/inbox-note/inbox-dismiss-confirmation-modal.tsx:55]: Nit pick but any reason why you don't just export it on the line its declared?

**Review by samueljseay — COMMENTED** [review]
Looking good, just some little things/suggestions from me.

File: packages/experimental/src/inbox-note/inbox-dismiss-confirmation-modal.tsx (hunk 13 above)
Comment (louwie17) [reply packages/experimental/src/inbox-note/inbox-dismiss-confirmation-modal.tsx:55]: No particular reason, will export it directly

**Comment by louwie17** [comment]
@samueljseay I updated it with your suggestions here: https://github.com/woocommerce/woocommerce-admin/pull/7006/commits/6a7602c7428f038e61da311ad55d6c6df21a7732
If you are able to give it another look :)

**Review by samueljseay — APPROVED** [review]
Fantastic stuff. Tests well too. Special kudos for such a high amount of quality tests too. You rock 🚀

**Pull request merged**